src/AppInstallerCommonCore/Public/winget/UserSettings.h

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#pragma once
+#include "AppInstallerStrings.h"
+
+#include <filesystem>
+#include <vector>
+#include <string>
+#include <variant>
+
+namespace AppInstaller::Settings
+{
+    // The type of argument.
+    enum class UserSettingsType
+    {
+        // Settings files don't exist. A file is created on the first call to the settings command.
+        Default,
+        // Loaded settings.json
+        Standard,
+        // Loaded settings.json.backup
+        Backup,
+    };
+
+    // The visual style of the progress bar.
+    enum class VisualStyle
+    {
+        NoVT,
+        Retro,
+        Accent,
+        Rainbow,
+    };
+
+    // Enum of settings.
+    // Must start at 0 to enable direct access to variant in UserSettings.
+    // Max must be last and unused.
+    enum class Setting : size_t
+    {
+        ProgressBarVisualStyle,
+        AutoUpdateTimeInMinutes,
+        Max
+    };
+
+    namespace details
+    {
+        template <Setting S>
+        struct SettingMapping
+        {
+            // json_t - type the setting in json.
+            // value_t - the type of this setting.
+            // DefaultValue - the value_t default value when setting is absent or semantically wrong.
+            // Path - json path to the property. See Json::Path in json.h for syntax. So far, this is sufficient
+            //        but since is "brief" and "untested" we might implement our own if needed.
+            // Validate - Function that does semantic validation.
+        };
+
+        template <>
+        struct SettingMapping<Setting::ProgressBarVisualStyle>
+        {
+            using json_t = std::string;
+            using value_t = VisualStyle;
+
+            static constexpr value_t DefaultValue = VisualStyle::Accent;
+            static constexpr std::string_view Path = ".visual.progressBar";
```

Comment:
>static constexpr std::string_view Path = ".visual.progressBar" [](start = 12, length = 62)

Do you actually get constexpr from this constructor?  It claims to, but I suppose I'm wary of that.  That's why I always use the string literal tag (sv).  But maybe it isn't necessary.

Reply:
I guess. Added just in case

---
In reply to: [443076363](https://github.com/microsoft/winget-cli/pull/436#discussion_r443076363) [](ancestors = 443076363)